### DISCUSSION TITLE
Help getting USB Coral TPU working on docker/unraid?

### DISCUSSION THREAD

**Discussion by shadow7412** [blakeblackshear/frigate#14935] [Ask A Question] [answered]
I'm having trouble getting frigate to detect my TPU.

> Attempting to load TPU as usb
> NoEdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.

The full error is exactly the same as [here](https://github.com/blakeblackshear/frigate/issues/7427).

So I did some investigation.

Here is an `lsusb` on the host before plugging it in.

    root@Lime:/mnt/user/steven/coral_example# lsusb
    Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 007 Device 003: ID 0781:5571 SanDisk Corp. Cruzer Fit
    Bus 007 Device 002: ID 1d6b:0104 Linux Foundation Multifunction Composite Gadget
    Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 005 Device 002: ID 0463:ffff MGE UPS Systems UPS
    Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 001 Device 003: ID 048d:8295 Integrated Technology Express, Inc. ITE Device(8595)
    Bus 001 Device 002: ID 8087:0a2b Intel Corp. Bluetooth wireless interface
    Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

 
And after (emphasis added to bring attention to the change);

    root@Lime:/mnt/user/steven/coral_example# lsusb
    Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 007 Device 003: ID 0781:5571 SanDisk Corp. Cruzer Fit
    Bus 007 Device 002: ID 1d6b:0104 Linux Foundation Multifunction Composite Gadget
    Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 006 Device 006: ID 1a6e:089a Global Unichip Corp.  <-----------------
    Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 005 Device 002: ID 0463:ffff MGE UPS Systems UPS
    Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
    Bus 001 Device 003: ID 048d:8295 Integrated Technology Express, Inc. ITE Device(8595)
    Bus 001 Device 002: ID 8087:0a2b Intel Corp. Bluetooth wireless interface
    Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

It coming up as that is apparently expected behaviour (according to https://github.com/google-coral/edgetpu/issues/536) and it's supposed to change vendor id after the first inference.

So, with that done, let's fire up a docker container so I can have a debian environment to run the examples. I'll just install some helpful stuff too.

    root@Lime:/mnt/user/steven/coral_example# docker run --rm -itv /dev/bus/usb:/dev/bus/usb debian
    root@f3024df7d7a6:/# apt-get update && apt-get install git curl usbutils gpg

    [OUTPUT TRUNCATED]

    done.
    root@f3024df7d7a6:/# lsusb

    root@f3024df7d7a6:/# lsusb
    Bus 008 Device 001: ID 1d6b:0003 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 007 Device 003: ID 0781:5571 SanDisk Cruzer Fit
    Bus 007 Device 002: ID 1d6b:0104 PiKVM Composite KVM Device
    Bus 007 Device 001: ID 1d6b:0002 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 006 Device 006: ID 1a6e:089a  <-----------------
    Bus 006 Device 001: ID 1d6b:0003 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 005 Device 002: ID 0463:ffff EATON Eaton 5P
    Bus 005 Device 001: ID 1d6b:0002 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 004 Device 001: ID 1d6b:0003 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 003 Device 001: ID 1d6b:0002 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 002 Device 001: ID 1d6b:0003 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller
    Bus 001 Device 003: ID 048d:8295 ITE Tech. Inc. ITE Device(8595)
    Bus 001 Device 002: ID 8087:0a2b   
    Bus 001 Device 001: ID 1d6b:0002 Linux 6.1.106-Unraid xhci-hcd xHCI Host Controller

 
 As you can see, the device is still there. It is interesting that the name has disappeared. I'm pretty sure this is the reason frigate can't detect the dongle. Has anyone else seen this?

Thanks.

**Comment by mimainka** [comment]
Usually the name of the coral changes after its been touched the first time.
So that your ID is staying 1a6e:089a indicates, it hasnt initialised yet.
I think I followed the following steps (but Im on a proxmox server) on the server itself and then it was initialised:
https://coral.ai/docs/accelerator/get-started/#runtime-on-linux

**Reply by shadow7412** [reply]
Is that something that needs to happen every reboot, or is it a one-time thing?

Also I, following those instructions inside a docker container forwarded in the same way, had the same issue anyway.

**Reply by mimainka** [reply]
It should happen just once.
And the execution should happen on host-level and then the initialised coral is passed to the container.

**Reply by shadow7412** [reply]
I had another go, using an OS mentioned in those docs rather than unraid. I managed to get the bird inference example working. So, the device isn't broken (which is good).

However, when I plugged it back into the server the device id was still coming up as 1a6e-etc... so does that mean it *isn't* a one-time thing? If that's the case I need to work out how I can reinitialize before frigate fires, given that the lack of TPU is enough to cause all of frigate to bomb out. (Not an ideal design decision for security camera software... :/)

So I guess I need to find a way to get that first inference to fire on the intended computer. It's somewhat of a shame that frigate itself isn't doing this...

**Reply by mimainka** [reply]
https://github.com/HarleyL7/unRAID-CoralTPU-Guide Maybe have a look at this. This could help initialising with Unraid.

**Reply by NickM-27** [reply]
Your understanding is incorrect, Frigate does initialize the USB coral on first inference, just like the examples

**Reply by mimainka** [reply]
Then my only guess would be a bad cable or not enough power from the usb port.

**Reply by shadow7412** [reply]
> Your understanding is incorrect, Frigate does initialize the USB coral on first inference, just like the examples

Good to hear. What about the differences between lsusb inside and outside of docket? Do you see the same thing on a working setup?

**Reply by NickM-27** [reply]
Should be `Google Inc` as the name, not sure why you don't see a name

**Reply by mimainka** [reply]
It should also have 
ID 18d1:9302

in the mentioned setup, the ID stays like the uninitialized one.

**Comment by towingie** [comment, marked as answer]
mine perfectly work with following compose options enabled:
```
services:
  figate:
      privileged: true
      ...
      - type: bind
        source: /dev/bus/usb
        target: /dev/bus/usb
```
not worked for me w/o them

**Reply by shadow7412** [reply]
I was under the impression that either the privileged flag or binding the usb directory should both have worked in isolation. It would seem I didn't try using both at the same time.

Thanks for your succinct example :)

**Reply by Red7Hunter** [reply]
Dumb question..


Where do I put these lines:

services:
  figate:
      privileged: true
      ...
      - type: bind
        source: /dev/bus/usb
        target: /dev/bus/usb


Sorry not sure where the services part goes?  Thanks!

**Reply by shadow7412** [reply]
@Red7Hunter it's a top level element of a compose.yaml file. Maybe this will help you: https://docs.docker.com/reference/compose-file/services/

**Reply by Red7Hunter** [reply]
Is there a way to add that detail in a unraid container?  That's where I think I'm stuck

**Reply by NickM-27** [reply]
you put it as a device passthrough, it should be in the community container by default

**Reply by Red7Hunter** [reply]
I assume this:
![coral](https://github.com/user-attachments/assets/29c4f82c-9611-4bba-933c-bc8571baed3c)

But I continue to get the errors like above

**Reply by NickM-27** [reply]
you should make your own support issue

**Reply by Red7Hunter** [reply]
I just finally figured it out... I had another model declared because I first was using CPU and doing openvino.. Once I commented that out it works!

**Reply by trevor4n** [reply]
@Red7Hunter where was the CPU model declared?
